### PR TITLE
Add support for Couchbase as a Vector Store

### PR DESCRIPTION
**Pull request by nithishr**
# Description

Add support for [Couchbase](https://couchbase.com/) as Vector Store. 

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/examples/CouchbaseVectorStoreDemo.ipynb
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/examples/CouchbaseVectorStoreDemo.ipynb:1]: Sorry for the confusion, can we actually move this to `docs/docs/examples/vector_stores/` ?

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+    """
+    Convert standard filters to Couchbase filter.
+
+    Args:
+        standard_filters (str): Standard Llama-index filters.
+
+    Returns:
+        Dictionary with Couchbase search filters.
+    """
+    if len(standard_filters.legacy_filters()) == 1:
```
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py:35]: I'm not sure if couchebase supports it, but `legacy_filters` means only `==` filters get used

In reality, we support a ton of filter arguments, and also `and` and `or` operators. Check out an integration like pinecone to see it in action. If couchbase supports this, would be nice to support it here too

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+    BasePydanticVectorStore,
+    VectorStoreQuery,
+    VectorStoreQueryResult,
+    MetadataFilters,
+)
+from llama_index.core.vector_stores.utils import (
+    node_to_metadata_dict,
+    metadata_dict_to_node,
+)
+from pydantic import PrivateAttr
```
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py:20]: lets use `from llama_index.core.bridge.pydantic import PrivateAttr` instead

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+        Delete a document by its reference document ID.
+
+        :param ref_doc_id: The reference document ID to be deleted.
+        :param **kwargs: Additional keyword arguments.
+        :type ref_doc_id: str
+        :return: None
+        :rtype: None
+        """
+        try:
+            self._collection.remove(ref_doc_id)
```
Comment (logan-markewich) [inline llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py:257]: this is actually inteded to delete any node that has `ref_doc_id == ref_doc_id` in its metadata (so a filtered delete) -- can we support that here?

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py (hunk 1 above)
Comment (nithishr) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py:35]: Couchbase supports filters as well but they are not the same filters as LlamaIndex.  For example, `>=` would have to be converted to a range query. 
So, I put in only the one filter that is the same. It is similar to Elastic. 
However, users can put in the couchbase specific filters via `cb_search_options`

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py (hunk 3 above)
Comment (nithishr) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py:257]: Thanks, I did not realize that. Will adapt it.

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py (hunk 1 above)
Comment (nithishr) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py]: I have found a way to support some more metadatafilters. I have added it in 2dc715400fd713a0dfc423d811647e62e3bc1bdc

**Review by logan-markewich — APPROVED** [review]

File: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py (hunk 1 above)
Comment (logan-markewich) [reply llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/llama_index/vector_stores/couchbase/base.py]: Awesome, thanks!

**Comment by nithishr** [comment]
Thank you @logan-markewich for the review. 
Can you help me to understand the errors in the check? 
```
Error: 1.20 [ERROR] 1 Exception encountered:

Engine traceback:
  in `tailor` goal

InvalidFieldException: llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/BUILD:1: Invalid field value for 'sources' in target llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase:llama-index-vector-stores-couchbase: Unmatched globs from llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase:llama-index-vector-stores-couchbase's `sources` field: ["llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/*.py", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/*.pyi"], excludes: ["llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/*_test.py", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/*_test.pyi", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/conftest.py", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/test_*.py", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/test_*.pyi", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/tests.py", "llama-index-integrations/vector_stores/llama-index-vector-stores-couchbase/tests.pyi"]
```

**Comment by logan-markewich** [comment]
Ah yea that's some pants stuff. Invalid build file. I can push fix to this PR

**Pull request merged**